shrine_with_cart/mdc_100_series/lib/short_bottom_sheet.dart

```diff
@@ -0,0 +1,578 @@
+      CurvedAnimation(
+          parent: _controller,
+          curve: _controller.status == AnimationStatus.forward
+              ? Interval(0.0, 0.25, curve: Curves.linear)
+              : Interval(0.25, 0.75, curve: Curves.linear)),
+    );
+
+    _cartOpacityAnimation = Tween<double>(begin: 0.0, end: 1.0).animate(
+      CurvedAnimation(
+          parent: _controller,
+          curve: _controller.status == AnimationStatus.forward
+              ? Interval(0.25, 0.75, curve: Curves.linear)
+              : Interval(0.75, 1.0, curve: Curves.linear)),
+    );
+
+    _shapeAnimation = TweenSequence(
+      <TweenSequenceItem<double>>[
+        TweenSequenceItem<double>(
+          tween: Tween<double>(begin: cornerRadius, end: cornerRadius * 0.4)
+              .chain(CurveTween(curve: _accelerateCurve)),
+          weight: 1.0 / 6.0,
+        ),
+        TweenSequenceItem<double>(
+          tween: Tween<double>(begin: cornerRadius * 0.4, end: 0.0)
+              .chain(CurveTween(curve: _decelerateCurve)),
+          weight: 5.0 / 6.0,
+        ),
+      ],
+    ).animate(
+      CurvedAnimation(
+          parent: _controller,
+          curve: _controller.status == AnimationStatus.forward
+              ? Interval(0.0, 0.35)
+              : Interval(0.17, 0.72)),
+    );
+  }
+
+  // Returns the correct width of the ShortBottomSheet based on the number of
+  // products in the cart.
+  double _getWidth(int numProducts) {
+    if (numProducts == 0) {
+      return 64.0;
+    } else if (numProducts == 1) {
+      return 136.0;
+    } else if (numProducts == 2) {
+      return 192.0;
+    } else if (numProducts == 3) {
+      return 248.0;
+    } else {
+      return 278.0;
+    }
+  }
+
+  // Updates _width based on the number of products in the cart.
+  void _updateWidth(int numProducts) {
+    _width = _getWidth(numProducts);
+  }
+
+  // Returns true if the cart is open and false otherwise.
+  bool get _isOpen {
+    final AnimationStatus status = _controller.status;
+    return status == AnimationStatus.completed ||
+        status == AnimationStatus.forward;
+  }
+
+  // Opens the ShortBottomSheet if it's open, otherwise does nothing.
+  void open() {
+    if (!_isOpen) {
+      setState(() {
+        _controller.forward();
+      });
+    }
+  }
+
+  // Closes the ShortBottomSheet if it's open, otherwise does nothing.
+  void close() {
+    if (_isOpen) {
+      setState(() {
+        _controller.reverse();
+      });
+    }
+  }
+
+  // Changes the padding between the left edge of the Material and the cart icon
+  // based on the number of products in the cart (padding increases when > 0
+  // products.)
+  void _adjustCartPadding(int numProducts) {
+    _cartPadding = numProducts == 0 ? 20.0 : 32.0;
+  }
+
+  bool _revealCart() {
+    if (_thumbnailOpacityAnimation.value == 0.0) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  Widget _buildThumbnails(int numProducts) {
+    return ExcludeSemantics(
```

Comment:
TODO: Say what `ExcludeSemantics` does in the codelab.

Reply:
Ok!